src/Storages/PartitionStrategy.cpp

```diff
@@ -0,0 +1,161 @@
+#include <Common/re2.h>
+#include <Storages/PartitionStrategy.h>
+#include <Parsers/ASTFunction.h>
+#include <Parsers/ASTLiteral.h>
+#include <Interpreters/TreeRewriter.h>
+#include <Interpreters/ExpressionAnalyzer.h>
+#include <Storages/PartitionedSink.h>
+#include <Functions/generateSnowflakeID.h>
+#include <Interpreters/Context.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+extern const int LOGICAL_ERROR;
+extern const int NOT_IMPLEMENTED;
+}
+
+
+namespace
+{
+    Names extractPartitionRequiredColumns(ASTPtr & partition_by, const Block & sample_block, ContextPtr context)
+    {
+        auto pby_clone = partition_by->clone();
+        auto syntax_result = TreeRewriter(context).analyze(pby_clone, sample_block.getNamesAndTypesList());
+        auto exp_analyzer = ExpressionAnalyzer(pby_clone, syntax_result, context).getActions(false);
+        return exp_analyzer->getRequiredColumns();
+    }
+
+    /*
+     * This isn't ideal, but I guess multiple formats can be specified and introduced.
+     * So I think it is simpler to keep it this way.
+     *
+     * Or perhaps implement something like `IInputFormat::getFileExtension()`
+     */
+    std::string formatToFileExtension(const std::string & format)
+    {
+        std::string lower_case_format;
```

Comment:
What is the difference with just `return Poco::toLower(format)`?

Reply:
I'll use Poco::toLower